File: solune/backend/src/models/agents.py
"""Pydantic models for the Agents section — Custom GitHub Agent CRUD."""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, Field


class AgentStatus(StrEnum):
    ACTIVE = "active"
    PENDING_PR = "pending_pr"
    PENDING_DELETION = "pending_deletion"
    IMPORTED = "imported"
    INSTALLED = "installed"


class AgentSource(StrEnum):
    LOCAL = "local"
    REPO = "repo"
    BOTH = "both"


class Agent(BaseModel):
    """API response model for repo-backed agent availability and agent workflow results."""

    id: str
    name: str
    slug: str
    description: str
    icon_name: str | None = None
    system_prompt: str = ""
    default_model_id: str = ""
    default_model_name: str = ""
    status: AgentStatus = AgentStatus.ACTIVE
    tools: list[str] = Field(default_factory=list)
    status_column: str | None = None
    github_issue_number: int | None = None
    github_pr_number: int | None = None
    branch_name: str | None = None
    source: AgentSource = AgentSource.LOCAL
    created_at: str | None = None
    agent_type: str = "custom"
    catalog_source_url: str | None = None
    catalog_agent_id: str | None = None
    imported_at: str | None = None


class AgentCreate(BaseModel):
    """Request body for creating a new agent."""

    name: str = Field(..., min_length=1, max_length=100)
    description: str = Field(default="", max_length=500)
    icon_name: str | None = Field(default=None, max_length=100)
    system_prompt: str = Field(..., min_length=1, max_length=30000)
    tools: list[str] = Field(default_factory=list)
    status_column: str = ""
    default_model_id: str = ""
    default_model_name: str = ""
    raw: bool = False  # If True, use exact content as-is without AI generation


class AgentUpdate(BaseModel):
    """Request body for updating an existing agent (P3)."""

    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, min_length=1, max_length=500)
    icon_name: str | None = Field(default=None, max_length=100)
    system_prompt: str | None = Field(default=None, min_length=1, max_length=30000)
    tools: list[str] | None = None
    default_model_id: str | None = None
    default_model_name: str | None = None


class AgentCreateResult(BaseModel):
    """Response for agent creation / update."""

    agent: Agent
    pr_url: str
    pr_number: int
    issue_number: int | None = None
    branch_name: str


class AgentDeleteResult(BaseModel):
    """Response for agent deletion."""

    success: bool
    pr_url: str
    pr_number: int
    issue_number: int | None = None


class AgentPendingCleanupResult(BaseModel):
    """Response for deleting stale pending agent rows from SQLite."""

    success: bool = True
    deleted_count: int


class AgentChatMessage(BaseModel):
    """Request body for chat refinement."""

    message: str
    session_id: str | None = None


class AgentChatResponse(BaseModel):
    """Response from chat refinement."""

    reply: str
    session_id: str
    is_complete: bool = False
    preview: AgentPreviewResponse | None = None


class AgentPreviewResponse(BaseModel):
    """Agent preview returned in chat responses."""

    name: str
    slug: str
    description: str
    icon_name: str | None = None
    system_prompt: str
    status_column: str = ""
    tools: list[str] = Field(default_factory=list)


class BulkModelUpdateRequest(BaseModel):
    """Request body for bulk model update — applies a target model to all agents."""

    target_model_id: str = Field(..., min_length=1, max_length=200)
    target_model_name: str = Field(..., min_length=1, max_length=200)


class BulkModelUpdateResult(BaseModel):
    """Response from bulk model update endpoint."""

    success: bool = True
    updated_count: int
    failed_count: int = 0
    updated_agents: list[str] = Field(default_factory=list)
    failed_agents: list[str] = Field(default_factory=list)
    target_model_id: str
    target_model_name: str


# ── Catalog / Import / Install Models ────────────────────────────────────


class CatalogAgent(BaseModel):
    """Agent listing from the Awesome Copilot catalog."""

    id: str
    name: str
    description: str
    source_url: str
    already_imported: bool = False


class ImportAgentRequest(BaseModel):
    """Request to import a catalog agent into a project."""

    catalog_agent_id: str
    name: str
    description: str
    source_url: str


class ImportAgentResult(BaseModel):
    """Response from importing a catalog agent."""

    agent: Agent
    message: str


class InstallAgentResult(BaseModel):
    """Response from installing an imported agent to a repository."""

    agent: Agent
    pr_url: str
    pr_number: int
    issue_number: int | None
    branch_name: str

File: solune/frontend/src/services/api/agents.ts
import type { PaginatedResponse } from '@/types';
import { request } from './client';

// ── Agent Types ─────────────────────────────────────────────────────────

export type AgentStatus = 'active' | 'pending_pr' | 'pending_deletion' | 'imported' | 'installed';
export type AgentSource = 'local' | 'repo' | 'both';

export interface AgentConfig {
  id: string;
  name: string;
  slug: string;
  description: string;
  icon_name: string | null;
  system_prompt: string;
  default_model_id: string;
  default_model_name: string;
  status: AgentStatus;
  tools: string[];
  status_column: string | null;
  github_issue_number: number | null;
  github_pr_number: number | null;
  branch_name: string | null;
  source: AgentSource;
  created_at: string | null;
  agent_type?: 'custom' | 'imported';
  catalog_source_url?: string | null;
  catalog_agent_id?: string | null;
  imported_at?: string | null;
}

export interface AgentCreate {
  name: string;
  description?: string;
  icon_name?: string | null;
  system_prompt: string;
  tools?: string[];
  status_column?: string;
  default_model_id?: string;
  default_model_name?: string;
  raw?: boolean;
}

export interface AgentUpdate {
  name?: string;
  description?: string;
  icon_name?: string | null;
  system_prompt?: string;
  tools?: string[];
  default_model_id?: string;
  default_model_name?: string;
}

export interface AgentCreateResult {
  agent: AgentConfig;
  pr_url: string;
  pr_number: number;
  issue_number: number | null;
  branch_name: string;
}

export interface AgentDeleteResult {
  success: boolean;
  pr_url: string;
  pr_number: number;
  issue_number: number | null;
}

export interface AgentPendingCleanupResult {
  success: boolean;
  deleted_count: number;
}

export interface AgentChatMessage {
  message: string;
  session_id?: string | null;
}

export interface AgentPreviewResponse {
  name: string;
  slug: string;
  description: string;
  system_prompt: string;
  status_column: string;
  tools: string[];
}

export interface AgentChatResponse {
  reply: string;
  session_id: string;
  is_complete: boolean;
  preview: AgentPreviewResponse | null;
}

export interface BulkModelUpdateResult {
  success: boolean;
  updated_count: number;
  failed_count: number;
  updated_agents: string[];
  failed_agents: string[];
  target_model_id: string;
  target_model_name: string;
}

export interface AgentMcpSyncResult {
  success: boolean;
  files_updated: number;
  files_skipped: number;
  files_unchanged: number;
  warnings: string[];
  errors: string[];
  synced_mcps: string[];
}

export interface CatalogAgent {
  id: string;
  name: string;
  description: string;
  source_url: string;
  already_imported: boolean;
}

export interface ImportAgentRequest {
  catalog_agent_id: string;
  name: string;
  description: string;
  source_url: string;
}

export interface ImportAgentResult {
  agent: AgentConfig;
  message: string;
}

export interface InstallAgentResult {
  agent: AgentConfig;
  pr_url: string;
  pr_number: number;
  issue_number: number | null;
  branch_name: string;
}

export const agentsApi = {
  list(projectId: string): Promise<AgentConfig[]> {
    return request<AgentConfig[]>(`/agents/${projectId}`);
  },

  listPaginated(
    projectId: string,
    params: { limit: number; cursor?: string },
  ): Promise<PaginatedResponse<AgentConfig>> {
    const qs = new URLSearchParams({ limit: String(params.limit) });
    if (params.cursor) qs.set('cursor', params.cursor);
    return request<PaginatedResponse<AgentConfig>>(`/agents/${projectId}?${qs}`);
  },

  pending(projectId: string): Promise<AgentConfig[]> {
    return request<AgentConfig[]>(`/agents/${projectId}/pending`);
  },

  clearPending(projectId: string): Promise<AgentPendingCleanupResult> {
    return request<AgentPendingCleanupResult>(`/agents/${projectId}/pending`, {
      method: 'DELETE',
    });
  },

  create(projectId: string, data: AgentCreate): Promise<AgentCreateResult> {
    return request<AgentCreateResult>(`/agents/${projectId}`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  update(projectId: string, agentId: string, data: AgentUpdate): Promise<AgentCreateResult> {
    return request<AgentCreateResult>(`/agents/${projectId}/${agentId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    });
  },

  delete(projectId: string, agentId: string): Promise<AgentDeleteResult> {
    return request<AgentDeleteResult>(`/agents/${projectId}/${agentId}`, {
      method: 'DELETE',
    });
  },

  chat(projectId: string, data: AgentChatMessage): Promise<AgentChatResponse> {
    return request<AgentChatResponse>(`/agents/${projectId}/chat`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  bulkUpdateModels(
    projectId: string,
    targetModelId: string,
    targetModelName: string
  ): Promise<BulkModelUpdateResult> {
    return request<BulkModelUpdateResult>(`/agents/${projectId}/bulk-model`, {
      method: 'PATCH',
      body: JSON.stringify({ target_model_id: targetModelId, target_model_name: targetModelName }),
    });
  },

  syncMcps(projectId: string): Promise<AgentMcpSyncResult> {
    return request<AgentMcpSyncResult>(`/agents/${projectId}/sync-mcps`, {
      method: 'POST',
    });
  },

  browseCatalog(projectId: string): Promise<CatalogAgent[]> {
    return request<CatalogAgent[]>(`/agents/${projectId}/catalog`);
  },

  importAgent(projectId: string, data: ImportAgentRequest): Promise<ImportAgentResult> {
    return request<ImportAgentResult>(`/agents/${projectId}/import`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  installAgent(projectId: string, agentId: string): Promise<InstallAgentResult> {
    return request<InstallAgentResult>(`/agents/${projectId}/${agentId}/install`, {
      method: 'POST',
    });
  },
};
